src/minimetrics/transport.py

```diff
@@ -21,30 +22,42 @@ class EncodingError(Exception):
 
 
 class RelayStatsdEncoder:
-    def encode(self, value: FlushedMetric) -> str:
+    def _encode(self, value: FlushedMetric, out: BytesIO):
+        _write = out.write
         (timestamp, metric_type, metric_name, metric_unit, metric_tags), metric = value
         metric_name = sanitize_value(metric_name) or "invalid-metric-name"
-        metric_values = ":".join(str(v) for v in metric.serialize_value())
-        serialized_metric_tags = self._get_metric_tags(metric_tags)
-        metric_tags_prefix = serialized_metric_tags and "|#" or ""
-        return f"{metric_name}@{metric_unit}:{metric_values}|{metric_type}{metric_tags_prefix}{serialized_metric_tags}|T{timestamp}"
-
-    def encode_multiple(self, values: Iterable[FlushedMetric]) -> str:
-        return "\n".join(self.encode(value) for value in values)
-
-    def _get_metric_tags(self, tags: MetricTagsInternal) -> str:
-        if not tags:
-            return ""
-
-        # We sanitize all the tag keys and tag values.
-        sanitized_tags = (
-            (sanitize_value(tag_key), sanitize_value(tag_value)) for tag_key, tag_value in tags
-        )
-
-        # We then convert all tags whose tag key is not empty to the string representation.
-        return ",".join(
-            f"{tag_key}:{tag_value}" for tag_key, tag_value in sanitized_tags if tag_key
-        )
+        _write(f"{metric_name}@{metric_unit}".encode())
+
+        for serialized_value in metric.serialize_value():
+            _write(b":")
+            _write(str(serialized_value).encode("utf-8"))
+
+        _write(f"|{metric_type}".encode("ascii"))
+
+        if metric_tags:
+            _write(b"|#")
+            first = True
+            for tag_key, tag_value in metric_tags:
+                tag_key = sanitize_value(tag_key)
+                if not tag_key:
+                    continue
+                if first:
+                    first = False
+                else:
+                    _write(b",")
+                _write(tag_key.encode("utf-8"))
+                _write(b":")
+                _write(sanitize_value(tag_value).encode("utf-8"))
+
+        _write(f"|T{timestamp}".encode("ascii"))
+
+    def encode_multiple(self, values: Iterable[FlushedMetric]) -> bytes:
+        out = BytesIO()
+        _write = out.write
+        for value in values:
+            self._encode(value, out)
+            _write(b"\n")
```

Comment:
Are we sure this `\n` at the end is fine? If we are at the last element

Reply:
Yeah